asdf/asdf.py

```diff
@@ -802,7 +764,21 @@ def _write_tree(self, tree, fd, pad_blocks):
         fd.write(b'\n')
 
         if len(tree):
-            yamlutil.dump_tree(tree, fd, self)
+            # yamlutil.custom_tree_to_tagged_tree looks for this attribute
+            # on the AsdfFile and if present, passes it to TypeIndex to
+            # record extensions used for conversion.
+            self._extensions_used = set()
```

Comment:
This is awful, but it's the way we currently handle state that only exists during serialization and I don't want to rock that particular boat right this moment.  I'd like to create a new class that gets passed around during serialization instead of the AsdfFile itself.